### PR TITLE
Fixed segfault

### PR DESCRIPTION
**Pull request by small-5**
Fix https://github.com/ntop/n2n/issues/719

### CODE REVIEW

**Comment by skyformat99** [comment]
> Fix #719

good

File: src/edge_utils.c [hunk 1]
```diff
@@ -3504,6 +3504,10 @@ int edge_conf_add_supernode (n2n_edge_conf_t *conf, const char *ip_and_port) {
     sock = (n2n_sock_t*)calloc(1,sizeof(n2n_sock_t));
     rv = supernode2sock(sock, ip_and_port);
 
+    while(rv == -2) {
```
Comment (Logan007) [inline src/edge_utils.c:3507]: This piece of code
- definitely needs some sleep, e.g. `sleep(2);` for not maxing out the CPU
- should go to another place (please see conversation comment)

File: src/n2n.c [hunk 2]
```diff
@@ -292,6 +292,7 @@ int supernode2sock (n2n_sock_t *sn, const n2n_sn_name_t addrIn) {
                     traceEvent(TRACE_WARNING, "supernode2sock fails to resolve supernode IPv4 address for %s", supernode_host);
                     rv = -1;
                 }
+                freeaddrinfo(ainfo); /* free everything allocated by getaddrinfo(). */
```
Comment (Logan007) [inline src/n2n.c:295]: Does this alone fix the segfault? Or does it only work when all supernodes are resolved?

**Review by Logan007 — COMMENTED** [review]
Thank you for your contribution!! :orange_heart: 

I am not too deep into the route setup code yet but I fully agree that errors should be reported, not by a segfault of course.

I think it generally is a good idea to have n2n output an error message by default if something fails, e.g. name resolution or route setup. Those errors may not be that severe to also require the edge to stop. Furthermore, I do not think that fully completed name resolution should be a common requirement to start-up the edge (some names might resolve way later, e.g. strategic reserve supernodes).

If your intent is to make the edge wait until **all supernode names resolve well**, we should do it somewhere around the edge's bootstrap using a new command line option (as this is optional behavior) to trigger the wait loop.

If however your intention is to make the edge wait until **the network is up**, I am not sure if this is the way to go. It might only work if names need to be resolved and might not work if plain IP addresses are given.

File: src/edge_utils.c (hunk 1 above)
Comment (small-5) [reply src/edge_utils.c:3507]: I originally wanted to join sleep 2 or even 5, but the test found that it was actually 5 seconds before the next parsing.
The purpose of adding this section is to loop the resolution when the domain name is set to be non IP and the resolution fails. If the program continues to run when the RV value is - 2, an error will be reported (sorry, I forgot the error prompt). After adding this section, this error can be avoided.

Comment (small-5) [reply src/edge_utils.c:3507]: I originally wanted to add sleep 2 or even 5, but the test found that there was already a 5 second interval for the next domain name resolution.
The purpose of adding this section is to loop the resolution when the domain name is set to be non IP and the resolution fails. If the program continues to run when the RV value is - 2, an error will be reported (sorry, I forgot the error prompt). After adding this section, this error can be avoided.

File: src/n2n.c (hunk 2 above)
Comment (small-5) [reply src/n2n.c:295]: Yes, this paragraph solves the segfault bug, and the previous paragraph solves a error output.

File: src/edge_utils.c (hunk 1 above)
Comment (Logan007) [reply src/edge_utils.c:3507]: The aim is to have the edge start-up even without the fully resolved names. It has already been working in my test scenarios. So, we need to find out more about this error, its triggers and maybe a solution that does not require to resolve all names during start-up.

Could we handle this error message in a separate issue / pull request, please?

**Review by Logan007 — CHANGES_REQUESTED** [review]
Please see my [comment](https://github.com/ntop/n2n/pull/720/files/c8ab0df834cd572b0d481fcb8e017f46387a825b#r667369771).

**Pull request closed without merging**